### PR TITLE
Add initial github action

### PR DESCRIPTION
**Pull request by seschis**
Implements the initial contrast scan github action.  Note that for PR, I'm mostly unsure if I'm writing idiomatic kotlin or not.  I'm still learning the KoTest Framework so there is nearly zero automated tests at the moment.  However I wanted to collect feedback on my kotlin code sooner rather tha later.  My intention is to keep adding tests while the review is occuring.

### CODE REVIEW

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            "https://api.github.com/repos",
+        )
+    }
+
+    private fun readEnvVar(name: String): String {
+        val v = System.getenv(name)
+        if (v == null) {
+            throw IllegalArgumentException("Environment var [$name] is not set!")
+        }
+        return v
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt:95]: you could write `System.getenv(name) ?: throw IllegalArgumentException(...)`

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    private fun writeSummaryToConsole(summary: ScanSummary, consoleLogger: Consumer<String?>) {
+        consoleLogger.accept("Scan completed")
+        consoleLogger.accept("New Results\t" + summary.totalNewResults())
+        consoleLogger.accept("Fixed Results\t" + summary.totalFixedResults())
+        consoleLogger.accept("Total Results\t" + summary.totalResults())
+    }
+
+    companion object {
+
+        @JvmStatic
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:177]: that is probably unnecessary here

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+        logger.info("Uploading " + artifact.fileName + " to Contrast Scan")
+
+        val codeArtifact: CodeArtifact
+        codeArtifact = try {
+            project.codeArtifacts().upload(artifact, prescan)
+        } catch (e: IOException) {
+            throw RuntimeException("Failed to upload code artifact to Contrast Scan", e)
+        } catch (e: HttpResponseException) {
+            throw RuntimeException("Failed to upload code artifact to Contrast Scan", e)
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:45]: all exceptions are unchecked in Kotlin, so there's no need to do this weird wrapping in RuntimeException we know from Java

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+class ContrastRemoteScanner(private val creds: ContrastPlatformCredentials, val apiUrl: URL, val projectName: String, val labelName: String) {
+    val sdk: ContrastSDK = connectToContrast(creds, apiUrl)
+
+    fun scan(prescan: Path, artifact: Path, scanOutputPath: Path, timeoutMs: Long) {
+        val project: Project = findOrCreateProject()
+
+        logger.info("Uploading " + artifact.fileName + " to Contrast Scan")
+
+        val codeArtifact: CodeArtifact
+        codeArtifact = try {
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:40]: you could initialize inline and let Kotlin infer the type: val codeArtifact = try...

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            "Contrast",
+            "static",
+            "ng",
+            "index.html#",
+            creds.orgId,
+            "scans",
+            projectId,
+            "scans",
+            scanId
+        )
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:100]: Why this roundabout way of concatenating the string? `val path = "/Contrast/static/ng/index.html#/${creds.orgId}/scans/${projectId}/scans/${scanId}`?

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+            )[timeoutMs, TimeUnit.MILLISECONDS]
+        } catch (e: ExecutionException) {
+            // try to unwrap the extraneous ExecutionException
+            val cause = e.cause
+            // ExecutionException should always have a cause, but its constructor does not enforce this,
+            // so check if the cause is null
+            val inner = cause ?: e
+            throw RuntimeException("Failed to retrieve Contrast Scan results", inner)
+        } catch (e: InterruptedException) {
+            Thread.currentThread().interrupt()
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:154]: why this call?

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        defaultValue = "https://teamserver-harmony-dev.contsec.com/Contrast",
+        description = ["Contrast Platform API url (\${DEFAULT-VALUE})"]
+    )
+    private lateinit var apiUrl: String
+
+    @CommandLine.Parameters(index= "0")
+    private lateinit var artifactPath: Path
+
+
+    @Throws(Exception::class)
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt:53]: is this necessary?

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            throw RuntimeException("Interrupted while retrieving Contrast Scan results", e)
+        } catch (e: TimeoutException) {
+            val duration = Duration.ofMillis(timeoutMs)
+            val durationString = if (duration.toMinutes() > 0) duration.toMinutes()
+                .toString() + " minutes" else (duration.toMillis() / 1000).toString() + " seconds"
+            throw RuntimeException(
+                "Failed to retrieve Contrast Scan results in $durationString", e
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:161]: where are these elaborately created exceptions evaluated?

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubCodeScaningAnalysis.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+            .setHeader("Authorization", authorization)
+            .POST(HttpRequest.BodyPublishers.ofString(
+                objectMapper.writeValueAsString(
+                    mapOf(
+                        "ref" to  gitHubScmData.ref,
+                        "commit_sha" to gitHubScmData.sha,
+                        "sarif" to b64Sarif,
+                        "checkout_uri" to "file:///github/workspace",
+                        "tool_name" to toolName,
+                    )
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubCodeScaningAnalysis.kt:72]: nice - this is idiomatic Kotlin :)

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubCodeScaningAnalysis.kt [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+                )
+            ))
+            .build()
+        var response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString())
+        logger.debug("response body: ${response.body()}")
+    }
+
+    companion object {
+        @JvmStatic
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubCodeScaningAnalysis.kt:81]: @JvmStatic is not necessary unless you are interacting with Java code

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+import java.nio.file.*
+import java.nio.file.attribute.BasicFileAttributes
+
+sealed class PreScanDataGenerator {
+    companion object {
+
+        @JvmStatic
+        private val logger = LoggerFactory.getLogger(javaClass.enclosingClass)
+
+        @Throws(Exception::class)
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt:18]: @Throws is not necessary while you are in the Kotlin universe. It adds metadata to the class files for interop with Java

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        fun generate(sourceDir: Path, outPath: Path) {
+            logger.debug("generating prescan data: source dir [$sourceDir],  output path [$outPath]")
+            val mapper = ObjectMapper()
+            val writer: Writer = FileWriter(outPath.toFile())
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt:22]: idiomatic Kotlin would let type inference play out here and live with `writer` being of type FileWriter

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                    .`with$schema`(URI.create("https://tbd/scan-input-metadata-schema-1.0.0.json"))
+                generator.writeStartObject()
+                generator.writeStringField("version", s.version.value())
+                generator.writeStringField("\$schema", s.`$schema`.toString())
+                generator.writeArrayFieldStart("paths")
+                // walk dir listing and stream out path data
+                Files.walkFileTree(
+                    sourceDir,
+                    object : SimpleFileVisitor<Path>() {
+                        @Throws(IOException::class)
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt:36]: @Throws should be unnecessary here

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+        private fun containsPath(path: Path, partial: Path): Boolean {
+            val iter: Iterator<Path> = path.toAbsolutePath().normalize().iterator()
+            while (iter.hasNext()) {
+                if (iter.next() == partial) {
+                    return true
+                }
+            }
+            return false
+        }
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt:78]: `private fun containsPath(path: Path, partial: Path) = p.normalize().any { it == partial }`

File: contrastscan-remote/src/test/java/com/contrastsecurity/scan/GitHubCodeScanningAnalysisTests.kt [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+import io.kotest.core.spec.style.BehaviorSpec
+import io.kotest.matchers.shouldBe
+
+class GitHubCodeScanningAnalysisTests : BehaviorSpec({
+    given("an incomplete environment") {
+        `when`("i call it") {
+            then("it should do this") {
+                1 + 2 shouldBe 3
+            }
+        }
```
Comment (arno-contrastsecurity) [inline contrastscan-remote/src/test/java/com/contrastsecurity/scan/GitHubCodeScanningAnalysisTests.kt:12]: I see you like spec-style unit tests :-)

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt (hunk 6 above)
Comment (seschis) [reply contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:154]: this was originally java code from our maven plugin that I copied and let the editor convert to Kotlin.
original code is: https://github.com/Contrast-Security-OSS/contrast-maven-plugin/blob/main/src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java#L304
I'll look more into the logic of this wait loop to see if there is a more idiomatic Kotlin-way to write it.

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt (hunk 8 above)
Comment (seschis) [reply contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt:161]: it was originally java code as a maven plugin ported to Kotlin.   Here is a link to the original source https://github.com/Contrast-Security-OSS/contrast-maven-plugin/blob/main/src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java#L347

I ended up changing this handler to `logger.error()` the detailed string, and then rethrow the original exception.

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt (hunk 7 above)
Comment (seschis) [reply contrastscan-remote/src/main/java/com/contrastsecurity/scan/GitHubActionCli.kt:53]: The editor added it for some reason.  So I just went along with it not really sure what it was saying.  I'll remove it.

File: contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt (hunk 14 above)
Comment (seschis) [reply contrastscan-remote/src/main/java/com/contrastsecurity/scan/PreScanDataGenerator.kt:78]: whoa!

File: contrastscan-remote/src/test/java/com/contrastsecurity/scan/GitHubCodeScanningAnalysisTests.kt (hunk 15 above)
Comment (seschis) [reply contrastscan-remote/src/test/java/com/contrastsecurity/scan/GitHubCodeScanningAnalysisTests.kt:12]: I wanted to try to write kotlin idiomatic tests and I was guessing the just using the JUnit4/5 libs and interfaces is not always the idiomatic way.   So I went looking for the standard kotlin test framework.  I came across the kotest framework which had some of these Spec-style testing language.  I've always liked thinking of tests in terms of expected "behavior" so I figured I'd give it a try.

**Pull request merged**